src/sys/unix/uds/datagram.rs

```diff
@@ -37,14 +37,31 @@ impl UnixDatagram {
         let fds = [-1; 2];
         let flags = libc::SOCK_DGRAM;
 
-        pair_descriptors(fds, flags)?;
-
-        Ok(unsafe {
-            (
-                UnixDatagram::from_raw_fd(fds[0]),
-                UnixDatagram::from_raw_fd(fds[1]),
-            )
-        })
+        #[cfg(not(any(target_os = "ios", target_os = "macos", target_os = "solaris")))]
+        {
+            pair_descriptors(fds, flags)?;
+            Ok(unsafe {
+                (
+                    UnixDatagram::from_raw_fd(fds[0]),
+                    UnixDatagram::from_raw_fd(fds[1]),
+                )
+            })
+        }
+
+        // Darwin and Solaris do not have SOCK_NONBLOCK or SOCK_CLOEXEC.
+        //
+        // In order to set those flags, additional `fcntl` sys calls must be
+        // made in `pair_descriptors` that are fallible. If a `fnctl` fails
+        // after the sockets have been created, the file descriptors will
+        // leak. Creating `s1` and `s2` below ensure that if there is an
+        // error, the file descriptors are closed.
+        #[cfg(any(target_os = "ios", target_os = "macos", target_os = "solaris"))]
+        {
```

Comment:
Why can't we use this code on other platforms as well?

Reply:
On other platforms, we'll know right away if `socket` succeeds or fails. That is the more common path of execution, and it saves on allocating extra `UnixDatagram` (or `UnixStreams`) that are not strictly necessary. I may just get rid of `pair_descriptors`. It's similar to why #1095 you had to allocate [here](https://github.com/Thomasdezeeuw/mio/blob/2ac1db79d9ddda7255e7edaf0be2bf1fbc31916b/src/sys/unix/mod.rs#L64) in the `fcntl` calls block, but not in the one [above](https://github.com/Thomasdezeeuw/mio/blob/2ac1db79d9ddda7255e7edaf0be2bf1fbc31916b/src/sys/unix/mod.rs#L58).

I may just remove `pair_descriptors` and have some code duplication between `UnixDatagram::pair` and `UnixStream::pair`.